### PR TITLE
Add initial implementation of OpenSergo Java client

### LINKED ISSUE #2: Add basic implementation of config subscriber
Add basic implementation of OpenSergo config subscriber in Java client.

### LINKED ISSUE #3: Add local config cache mechanism
Add local cache mechanism for received OpenSergo low-level config.

### LINKED ISSUE #4: Add gRPC bi-directional transport layer in OpenSergo client
Add gRPC bi-directional transport layer in OpenSergo client. The client should follow the *OpenSergo Universal Transport Service*.

### PR DESCRIPTION
**Pull request by sczyh30**
Add initial implementation of OpenSergo Java client.

* Add fundamental client subscribe and config cache mechanism
* Add fundamental implementation of OpenSergo gRPC client

### CODE REVIEW

**Comment by sczyh30** [comment]
An example of how data-plane integrates with OpenSergo Java SDK:

```java
OpenSergoClient client = new OpenSergoClient(host, port);
client.start();

// Push-model
client.subscribeConfig(new SubscribeKey(namespace, appName, configKind),
    new OpenSergoConfigSubscriber() {
        @Override
        public boolean onConfigUpdate(SubscribeKey subscribeKey, Object dataList) {
            // Handle received config here
            System.out.println("key: " + subscribeKey + ", data: " + dataList);
            return true;
        }
    });
```

File: src/main/proto/fault_tolerance/v1/fault_tolerance.proto [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  FaultToleranceActionRef action = 3;
+}
+
+// RateLimitStrategy
+
+message RateLimitStrategy {
+  string name= 1;
+
+  string metric_type = 2;
+  string limit_mode = 3;
```
Comment (jnan806) [inline src/main/proto/fault_tolerance/v1/fault_tolerance.proto]: recommended to define enum

File: src/main/proto/fault_tolerance/v1/fault_tolerance.proto [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  int64 threshold = 4;
+  int32 stat_duration = 5;
+  io.opensergo.proto.common.v1.TimeUnit stat_duration_time_unit = 6;
+}
+
+// ConcurrencyLimitStrategy
+
+message ConcurrencyLimitStrategy {
+  string limit_mode = 1;
```
Comment (jnan806) [inline src/main/proto/fault_tolerance/v1/fault_tolerance.proto]: recommended to use the same enum with `limit_mode` in `RateLimitStrategy`

File: src/main/proto/service_contract/v1/service_contract.proto [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+  // ReportMetadata report Metadata to server
+  rpc ReportMetadata(ReportMetadataRequest) returns (ReportMetadataReply);
+}
+
+message ReportMetadataRequest {
+  string app_name = 1;
+  Node node = 2;
+
+  repeated ServiceMetadata service_metadata = 3;
+  reserved 4 to 16;
```
Comment (jnan806) [inline src/main/proto/service_contract/v1/service_contract.proto:37]: reserved tag conflict with docs https://github.com/opensergo/opensergo-specification/blob/main/specification/zh-Hans/README.md
maybe the docs is needed to change the tag range

image

File: src/main/proto/fault_tolerance/v1/fault_tolerance.proto [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+message CircuitBreakerSlowCondition {
+  int32 max_allowed_rt_millis = 1;
+}
+
+message CircuitBreakerErrorCondition {
+
+}
+
+message CircuitBreakerStrategy {
+  string strategy = 1;
```
Comment (jnan806) [inline src/main/proto/fault_tolerance/v1/fault_tolerance.proto]: recommended to define an enum

**Review by jnan806 — CHANGES_REQUESTED** [review]

File: src/main/proto/fault_tolerance/v1/fault_tolerance.proto (hunk 2 above)
Comment (sczyh30) [reply src/main/proto/fault_tolerance/v1/fault_tolerance.proto]: Good idea.

**Comment by jnan806** [comment]
是否应该将 `*.proto` 文件抽取到 [opensergo/opensergo-proto](https://github.com/opensergo/opensergo-proto) 项目

**Comment by sczyh30** [comment]
> 是否应该将 `*.proto` 文件抽取到 [opensergo/opensergo-proto](https://github.com/opensergo/opensergo-proto) 项目

对应的 proto 会提交到 opensergo-proto 项目里面；不过如何在 SDK 项目中比较方便地引用 proto，社区可以给一些更好的建议

**Comment by jnan806** [comment]
refer to [opensergo-proto` 的 [#issue 3] (https://github.com/opensergo/opensergo-proto/issues/3)

**Comment by jnan806** [comment]
> 
> 对应的 proto 会提交到 opensergo-proto 项目里面；不过如何在 SDK 项目中比较方便地引用 proto，社区可以给一些更好的建议

refer to https://github.com/opensergo/opensergo-proto/issues/3

**Comment by jnan806** [comment]
@sczyh30

PTAL！THX.

针对 issue https://github.com/opensergo/opensergo-proto/issues/3

已经将 `grpc协议`的`java`和`go`两中SDK的依赖实现提交在 PR 中，麻烦review并讨论下是否合理。

https://github.com/opensergo/opensergo-proto/pull/4

File: src/main/proto/fault_tolerance/v1/fault_tolerance.proto (hunk 1 above)
Comment (sczyh30) [reply src/main/proto/fault_tolerance/v1/fault_tolerance.proto]: Updated.

**Comment by sczyh30** [comment]
Hi, I've updated the latest proto and polished SDK code (a lot of bugs has been fixed...).

cc @opensergo/opensergo-maintainer

**Comment by jnan806** [comment]
ok, I've synchronized my fork branch https://github.com/jnan806/opensergo-java-sdk-review-sczyh30/tree/stubcode-based-sczyh30

**Comment by jnan806** [comment]
ok, I've synchronized my fork branch https://github.com/jnan806/opensergo-java-sdk-review-sczyh30/tree/stubcode-based-sczyh30

**Comment by jnan806** [comment]
ok, I've synchronized my fork branch https://github.com/jnan806/opensergo-java-sdk-review-sczyh30/tree/stubcode-based-sczyh30   
Please take a review

**Pull request merged**